python/ql/src/experimental/semmle/python/security/dataflow/XmlInjectionCustomizations.qll

```diff
@@ -0,0 +1,95 @@
+/**
+ * Provides default sources, sinks and sanitizers for detecting
+ * "ldap injection"
+ * vulnerabilities, as well as extension points for adding your own.
+ */
+
+private import python
+private import semmle.python.dataflow.new.DataFlow
+private import experimental.semmle.python.Concepts
+private import semmle.python.dataflow.new.RemoteFlowSources
+private import semmle.python.dataflow.new.BarrierGuards
+private import semmle.python.ApiGraphs
+
+/**
+ * Provides default sources, sinks and sanitizers for detecting "xml injection"
+ * vulnerabilities, as well as extension points for adding your own.
+ */
+module XmlInjection {
+  /**
+   * A data flow source for "xml injection" vulnerabilities.
+   */
+  abstract class Source extends DataFlow::Node { }
+
+  /**
+   * A data flow sink for "xml injection" vulnerabilities.
+   */
+  abstract class Sink extends DataFlow::Node { }
+
+  /**
+   * A sanitizer guard for "xml injection" vulnerabilities.
+   */
+  abstract class SanitizerGuard extends DataFlow::BarrierGuard { }
+
+  /**
+   * A data flow sink for XML parsing libraries.
+   *
+   * See `XML::XMLParsing`.
+   */
+  abstract class XMLParsingSink extends Sink { }
+
+  /**
+   * A data flow sink for XML parsers.
+   *
+   * See `XML::XMLParser`
+   */
+  abstract class XMLParserSink extends Sink { }
+
+  /**
+   * A source of remote user input, considered as a flow source.
+   */
+  class RemoteFlowSourceAsSource extends Source, RemoteFlowSource { }
+
+  /**
+   * An xml parsing operation, considered as a flow sink.
+   */
+  class XMLParsingInputAsSink extends XMLParsingSink {
+    XMLParsingInputAsSink() { this = any(XML::XMLParsing xmlParsing).getAnInput() }
+  }
+
+  /**
+   * An xml parsing operation vulnerable to `kind`.
+   */
+  predicate xmlParsingInputAsVulnerableSink(DataFlow::Node sink, string kind) {
+    exists(XML::XMLParsing xmlParsing |
+      sink = xmlParsing.getAnInput() and xmlParsing.vulnerable(kind)
+    )
+  }
```

Comment:
This should use `XMLParsingInputAsSink` instead of `XML::XMLParsing` to match inheritance but that way `XML::XMLParsing`'s `vulnerable` predicate would not not be shared :(